internal/rotate_config.go

```diff
@@ -0,0 +1,165 @@
+package internal
+
+import (
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"log"
+
+	"github.com/ThalesIgnite/crypto11"
+)
+
+type fullCfgStep struct{}
+
+func (s fullCfgStep) Name() string {
+	return "Update local configuration with new key"
+}
+
+func (s fullCfgStep) Execute(handler *CertRotationHandler) error {
+	crypto, err := getCryptoHandler(handler)
+	if err != nil {
+		return err
+	}
+	defer crypto.Close()
+
+	// Open/decrypt full config with current key
+	config, err := UnmarshallFile(handler.crypto, handler.app.EncryptedConfig, true)
+	if err != nil {
+		return fmt.Errorf("Unable open current encrypted config: %w", err)
+	}
+
+	// Encrypt with new key
+	cfgBytes, err := encryptConfig(crypto, config)
+	if err != nil {
+		return err
+	}
+	handler.State.FullConfigEncrypted = string(cfgBytes)
+	return nil
+}
+
+type deviceCfgStep struct{}
+
+func (s deviceCfgStep) Name() string {
+	return "Update device specific configuration on server with new key"
+}
+
+func (s deviceCfgStep) Execute(handler *CertRotationHandler) error {
```

Comment:
I think we assume that a user cannot add a device config (fioctl) while the cert rotation is in progress. In this case a race condition may happen, the user's device config encrypted by an old pub key may arrive to ota-lite just after this code pushed config encrypted by a new key.

Reply:
correct. and this new logic keeps the race condition to a very small window. We could only have a problem if:
1. device asks for its config
2. user updates device's config
3. device re-encrypts and uploads new config

The time between steps 1 and 3 is tiny. Once 3 has been performed the device-gateway can flag that device's config as "not updateable" and it can stay this way until the device finally checks in with the new key.

Reply:
ok, the use-case you provided seems restorable/fixable.
I am not sure that the case I mentioned can be fixed since we end up with a device config encrypted by an old key that has been already rotated.

Reply:
@doanac I think that when the user requests to rotate the device key, we may immediately freeze any further config updates until the devices informs about the successful rotation. IMHO the ota-lite has the ability to be enough intelligent on this.

Then any potential window for even a hypothetic race condition like this is closed.

Reply:
Furthermore, when the user comes up and says smth like "_I screwed up my device which is now frozen and need to update it_" - then we can still help them via the customer support channel.

Reply:
that's fine. we can work through that when we get to device-gateway and fioctl work